### PR TITLE
fix(pipelines): undeployable due to dependency cycle

### LINKED ISSUE #18673: CDK Pipelines: Circular dependency on initial CDK Deploy
### What is the problem?

I am trying to deploy a pipeline using the aws-cdk/pipelines library and I am receiving a circular dependency error. I have no other resources in this stack other than the pipeline itself.

### Reproduction Steps

Outer Stack: 
```
import 'source-map-support/register';
import * as cdk from 'aws-cdk-lib';
import { SplitMonitorStackStack } from '../lib/MainPipelineStack';

const app = new cdk.App();
new SplitMonitorStackStack(app, 'SplitMonitorMainStack');
app.synth();
```
Inner Stack:

```

import { Stack, StackProps,SecretValue} from 'aws-cdk-lib';
import { Construct } from 'constructs'; // construct is in a separate package
import * as pipelines from 'aws-cdk-lib/pipelines';

export class SplitMonitorStackStack extends Stack {
  constructor(scope: Construct, id: string, props?: StackProps) {
    super(scope, id, props);


    const splitMonitoringStackPipeline = new pipelines.CodePipeline(this, "pipeline", {
      pipelineName: "Split-Monitor-Main",
      synth: new pipelines.ShellStep('Synth', {
          input: pipelines.CodePipelineSource.gitHub('PlaceholderRepoName', "main", {
            authentication: SecretValue.secretsManager('Token'),
          }),
          commands: [
            'npm install -g aws-cdk@latest',
            'npm ci',
            'npm run build',
            'npx cdk synth',
          ],
        }),
        crossAccountKeys: true
    });

  }
}

```

### What did you expect to happen?

The pipeline and CDK deploy do be successful. 

### What actually happened?

`
SplitMonitorMainStack failed: Error [ValidationError]: Circular dependency between resources: [pipelineUpdatePipelineSelfMutation14A96D2F, pipelinePipelineBuildSynthCdkBuildProject4237770A, pipelinePipelineBuildSynthCdkBuildProjectRoleDefaultPolicyB3981181, pipelineUpdatePipelineSelfMutationRoleDefaultPolicy095404B8, pipelinePipeline4163A4B1, pipelinePipelineUpdatePipelineSelfMutateCodePipelineActionRoleDefaultPolicy09E3C254, pipelinePipelineSourceIvellaSplitMonitorStackWebhookResource4158C29A, pipelinePipelineBuildSynthCodePipelineActionRoleDefaultPolicy7B8E3FCA]
    at Request.extractError (/opt/homebrew/lib/node_modules/aws-cdk/node_modules/aws-sdk/lib/protocol/query.js:50:29)
    at Request.callListeners (/opt/homebrew/lib/node_modules/aws-cdk/node_modules/aws-sdk/lib/sequential_executor.js:106:20)
    at Request.emit (/opt/homebrew/lib/node_modules/aws-cdk/node_modules/aws-sdk/lib/sequential_executor.js:78:10)
    at Request.emit (/opt/homebrew/lib/node_modules/aws-cdk/node_modules/aws-sdk/lib/request.js:686:14)
    at Request.transition (/opt/homebrew/lib/node_modules/aws-cdk/node_modules/aws-sdk/lib/request.js:22:10)
    at AcceptorStateMachine.runTo (/opt/homebrew/lib/node_modules/aws-cdk/node_modules/aws-sdk/lib/state_machine.js:14:12)
    at /opt/homebrew/lib/node_modules/aws-cdk/node_modules/aws-sdk/lib/state_machine.js:26:10
    at Request.<anonymous> (/opt/homebrew/lib/node_modules/aws-cdk/node_modules/aws-sdk/lib/request.js:38:9)
    at Request.<anonymous> (/opt/homebrew/lib/node_modules/aws-cdk/node_modules/aws-sdk/lib/request.js:688:12)
    at Request.callListeners (/opt/homebrew/lib/node_modules/aws-cdk/node_modules/aws-sdk/lib/sequential_executor.js:116:18) {
  code: 'ValidationError',
  time: 2022-01-26T18:54:30.449Z,
  requestId: '913f9618-4a60-483a-b8b4-aec4be38998a',
  statusCode: 400,
  retryable: false,
  retryDelay: 755.4424894449671
}
Circular dependency between resources: [pipelineUpdatePipelineSelfMutation14A96D2F, pipelinePipelineBuildSynthCdkBuildProject4237770A, pipelinePipelineBuildSynthCdkBuildProjectRoleDefaultPolicyB3981181, pipelineUpdatePipelineSelfMutationRoleDefaultPolicy095404B8, pipelinePipeline4163A4B1, pipelinePipelineUpdatePipelineSelfMutateCodePipelineActionRoleDefaultPolicy09E3C254, pipelinePipelineSourceIvellaSplitMonitorStackWebhookResource4158C29A, pipelinePipelineBuildSynthCodePipelineActionRoleDefaultPolicy7B8E3FC
`

### CDK CLI Version

2.9.0

### Framework Version

_No response_

### Node.js Version

v16.13.0

### OS

macOS Monterey

### Language

Typescript

### Language Version

_No response_

### Other information

_No response_

### PR DESCRIPTION
**Pull request by rix0rrr**
A dependency cycle was inadvertently introduced to CDK Pipelines in #18492.

Fix that dependency cycle, and also one in Cognito IdentityPools.

Add facilities to the `assertions` library to automatically detect this in the future,
to stop errors like this from slipping in.

We could make it a separate assertion method
(`Template.fromStack().assertNoCycles()`), but the only thing that will
do is give you an opportunity to forget to put the test in.

Instead, we just check it by default for every generated template.

Fixes #18673.

----

*By submitting this pull request, I confirm that my contribution is made under the terms of the Apache-2.0 license*

### CODE REVIEW

File: packages/@aws-cdk/assertions/test/template.test.ts [hunk 1]
```diff
@@ -1084,6 +1084,25 @@ describe('Template', () => {
       expect(Object.keys(result).length).toEqual(0);
     });
   });
+
+  test('throws when given a templat with cyclic dependencies', () => {
```
Comment (LukvonStrom) [inline packages/@aws-cdk/assertions/test/template.test.ts:1088]: *template

**Review by njlynch — APPROVED** [review]

File: packages/@aws-cdk/aws-cognito-identitypool/lib/identitypool.ts [hunk 2]
```diff
@@ -437,7 +437,12 @@ export class IdentityPool extends Resource implements IIdentityPool {
       unauthenticatedRole: this.unauthenticatedRole,
       roleMappings: props.roleMappings,
     });
-    attachment.node.addDependency(this);
+
+    // This added by the original author, but it's causing cyclic dependencies.
+    // Don't know why this was added in the first place, but I'm disabling it for now and if
+    // no complaints come from this, we're probably safe to remove it altogether.
+    // attachment.node.addDependency(this);
+    Array.isArray(attachment);
```
Comment (LukvonStrom) [inline packages/@aws-cdk/aws-cognito-identitypool/lib/identitypool.ts:445]: what does this line do?

Comment (rix0rrr) [reply packages/@aws-cdk/aws-cognito-identitypool/lib/identitypool.ts:445]: It prevents an "unused variable" error :)

File: packages/@aws-cdk/assertions/lib/private/cyclic.ts [hunk 3]
```diff
@@ -0,0 +1,175 @@
+import { Resource, Template } from './template';
+
+/**
+ * Check a template for cyclic dependencies
+ *
+ * This will make sure that we don't happily validate templates
+ * in unit tests that wouldn't deploy to CloudFormation anyway.
+ */
+export function checkTemplateForCyclicDependencies(template: Template): void {
```
Comment (madeline-k) [inline packages/@aws-cdk/assertions/lib/private/cyclic.ts:9]: So cool! 🎉

**Pull request merged**